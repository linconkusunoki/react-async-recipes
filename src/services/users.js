import api from './api'

export async function fetchUsers() {
  try {
    const resp = await api.get('/users')
    return resp
  } catch (error) {
    return error
  }
}
