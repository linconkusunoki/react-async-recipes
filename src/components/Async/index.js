import React from 'react'
import { fetchUsers } from '../../services/users'
import useFetch from '../../hooks/useFetch'

function Async() {
  const users = useFetch(fetchUsers)

  if (users.isLoading) {
    return <p>Loading Users...</p>
  }

  if (users.error) {
    return <p>Error while loading users!</p>
  }

  return (
    <ul>
      {users.response.data.map(user => (
        <li key={user.id}>{user.firstName}</li>
      ))}
    </ul>
  )
}

export default Async
