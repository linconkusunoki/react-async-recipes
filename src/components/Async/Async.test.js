import React from 'react'
import { render, wait } from '@testing-library/react'
import Async from './'
import { fetchUsers } from '../../services/users'

jest.mock('../../services/users')

it('should render a list of users from api', async () => {
  const users = [{ id: 1, firstName: 'john' }]
  fetchUsers.mockResolvedValueOnce({ data: users })

  const { getByText, container } = render(<Async />)

  expect(getByText('Loading Users...'))
  await wait(() => expect(getByText('john')).toBeTruthy())
  expect(fetchUsers).toHaveBeenCalled()
  expect(container.firstChild).toMatchSnapshot()
})
