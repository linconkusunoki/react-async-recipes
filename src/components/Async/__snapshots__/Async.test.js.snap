// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a list of users from api 1`] = `
<ul>
  <li>
    john
  </li>
</ul>
`;
